onnxruntime/core/providers/dnnl/dnnl_execution_provider.h

```diff
@@ -94,6 +114,13 @@ class DNNLExecutionProvider : public Provider_IExecutionProvider {
   std::vector<IAllocatorUniquePtr<void>> biass_buffers_;
   OrtMutex mutex_;
 
+#ifdef ENABLE_TRAINING
+  // map used to hold and lookup forward DnnlKernels. This should only be needed in when
+  // running in training mode.The backward Kernels need access the forward kernals; typically
+  // to obtain the forward primitive description but it may be need for other items like
+  // accessing workspace memory.
+  std::map<onnxruntime::NodeIndex, std::shared_ptr<ort_dnnl::DnnlKernel>> fwd_kernal_map_;
```

Comment:
1. verified that a new DnnlExecutionProvider is created per session by creating multiple sessions in the mnist sample.
2. In testing the fwd_kernel_map_ is always accessed by the same thread the entire time the training is running.   The only time it is accessed by a different thread is at the end of the training session when it runs an inference session to check the accuracy of the trained graph. Since they are run sequentially there are never two threads trying to access the map at the same time.